seaworthy/logs.py

```diff
@@ -82,32 +82,79 @@ def wait_for_logs_matching(container, matcher, timeout=10, encoding='utf-8',
         matcher, _last_few_log_lines(container)))
 
 
-class SequentialLinesMatcher(object):
+class LogMatcher(ABC):
+    """
+    Abstract base class for log matchers.
     """
-    Matcher that takes a list of matchers, and uses one after the next after
-    each has a successful match. Returns True ("matches") on the final match.
 
-    **Note:** This is a *stateful* matcher. Once it has done its matching,
-    you'll need to create a new instance.
+    @abstractmethod
+    def match(self, log_line):
+        """
+        Return ``True`` if the matcher matches a line, otherwise ``False``.
+        """
+
+    @abstractmethod
+    def args_str(self):
+        """
+        Return an args string for the repr.
+        """
+
+    def __call__(self, log_line):
+        return self.match(log_line)
+
+    def __str__(self):
+        return '{}({})'.format(type(self).__name__, self.args_str())
+
+    def __repr__(self):
+        return str(self)
+
+
+def to_matcher(matcher_factory, obj):
+    return obj if isinstance(obj, LogMatcher) else matcher_factory(obj)
+
+
+class CombinationLogMatcher(LogMatcher):
+    """
+    Matcher that combines multiple input matchers.
     """
     def __init__(self, *matchers):
         self._matchers = matchers
-        self._position = 0
 
     @classmethod
-    def by_equality(cls, *rhs):
-        return SequentialLinesMatcher(*map(EqualsMatcher, rhs))
+    def by_equality(cls, *expected_lines):
+        """
+        Construct an instance of this combination matcher from a list of
+        expected log lines and/or LogMatcher instances.
+        """
+        return cls(*(to_matcher(EqualsMatcher, l) for l in expected_lines))
 
     @classmethod
     def by_regex(cls, *patterns):
-        return SequentialLinesMatcher(*map(RegexMatcher, patterns))
+        """
+        Construct an instance of this combination matcher from a list of
+        regex patterns and/or LogMatcher instances.
+        """
+        return cls(*(to_matcher(RegexMatcher, p) for p in patterns))
+
+
+class OrderedLinesMatcher(CombinationLogMatcher):
+    """
+    Matcher that takes a list of matchers, and uses one after the next after
+    each has a successful match. Returns True ("matches") on the final match.
 
-    def __call__(self, line):
+    **Note:** This is a *stateful* matcher. Once it has done its matching,
+    you'll need to create a new instance.
+    """
+    def __init__(self, *matchers):
+        super().__init__(*matchers)
+        self._position = 0
```

Comment:
Perhaps we should do matched/unmatched lists here like the unordered matcher? Is it possible to share more logic, or not worth the effort? At one point I did do the two lists that with `SequentialLinesMatcher` but then changed my mind...

Reply:
It's certainly possible, but I don't really want to bake too many assumptions into the abstraction and I didn't think it was worth the effort of changing how this implementation worked if I wasn't going to factor out the common bits.

I also had some vague thoughts of adding a `.reset()` method to clear state and allow the matcher to be reused, but eventually decided against that because I couldn't come up with a real use case.